### PR TITLE
[DOC] Add transactions how-to and example

### LINKED ISSUE #4301: Add more detail on transaction semantics
We have a [broad overview](https://docs.pachyderm.com/deploy-manage/manage/batching_pachyderm_with_transactions/) of transactions in the docs, but don't cover much detail. In particular:

* Can transactions be nested, or are they related to each other in any way? (They aren't.)
* What is happening at the API level? (Similar to auth tokens, transaction IDs get passed as gRPC metadata.)
* How are conflicts between transactions resolved, if at all? (I'm not sure of the answer to this one myself, but my guess is that the second transaction that's committed will fail.)

### PR DESCRIPTION
**Pull request by svekars**
Closes #4301

### CODE REVIEW

File: doc/docs/master/deploy-manage/manage/use-transactions-to-run-multiple-commands.md [hunk 1]
```diff
@@ -0,0 +1,221 @@
...
+             "source": 3,
+             "active_transaction": "7a81eab5-e6c6-430a-a5c0-1deb06852ca5",
+             "cluster_name": "minikube",
+             "auth_info": "minikube",
+             "namespace": "default"
+           },
+    ```
+
+After you start a transaction, you can add supported commands, such
+as create a repo, create a branch, and so on, to th transaction. You
```
Comment (JoeyZwicker) [inline doc/docs/master/deploy-manage/manage/use-transactions-to-run-multiple-commands.md]:
```suggestion
as `create branch`, `start commit`, and so on, to the transaction. You
```

Comment (JoeyZwicker) [reply doc/docs/master/deploy-manage/manage/use-transactions-to-run-multiple-commands.md]: remove: 

"you can also start a commit within a transaction"

File: doc/docs/master/deploy-manage/manage/use-transactions-to-run-multiple-commands.md [hunk 2]
```diff
@@ -0,0 +1,221 @@
...
+can also start a commit within a transaction. If you close a transaction
+before you close a commit, the commands that you have specified inside
+the transaction and inside the
+commit will not be applied until you close all commits that you have
+started in that transaction.
```
Comment (JoeyZwicker) [inline doc/docs/master/deploy-manage/manage/use-transactions-to-run-multiple-commands.md]: > If you close a transaction
before you close a commit, the commands that you have specified inside
the transaction and inside the
commit will not be applied until you close all commits that you have
started in that transaction.

I dont really understand what this is saying and I dont think it's accurate. There's no special logic around waiting on commits or something with transactions. Commits have the property that they dont do anything until you close them. 

Perhaps a better way to describe this is:

"All commands that are performed in a transaction are queued up and not executed against the actual cluster. One you finish the transaction, all queued command are executed atomically."

File: doc/docs/master/deploy-manage/manage/use-transactions-to-run-multiple-commands.md [hunk 3]
```diff
@@ -0,0 +1,221 @@
...
+create repo
+delete repo
+start commit
+finish commit
+delete commit
+create branch
+delete branch
+```
+
+Each time you ad a command to a transaction, Pachyderm validates the
```
Comment (JoeyZwicker) [inline doc/docs/master/deploy-manage/manage/use-transactions-to-run-multiple-commands.md]:
```suggestion
Each time you add a command to a transaction, Pachyderm validates the
```

File: examples/transactions/README.md [hunk 4]
```diff
@@ -0,0 +1,236 @@
...
+   ID                               PIPELINE STARTED      DURATION RESTART PROGRESS  DL UL STATE
+   ```
+
+1. Add the data to the `raw_data` repository:
+
+   ```bash
+   $ pachctl put file raw_data@master:iris.csv -f noisy_iris.csv
+   ```
+
+1. Verify that the `iris.csv` file is added to the repository:
```
[the remark this answers is not in the document]
Comment (svekars) [reply examples/transactions/README.md]: If the purpose of an example is to teach users how to use Pachyderm, then yes, I think we need this. When I use something for the first time, how do I know that the command that I've run took effect? These are the people who are new to Pachyderm.

File: examples/transactions/README.md [hunk 5]
```diff
@@ -0,0 +1,236 @@
...
+   ```
+
+1. Close the commit to the `raw_data` repository that you have
+   started within the transaction:
+
+   ```bash
+   $ pachctl finish commit raw_data@master
+   ```
+
+   Still no jobs run for the `model` pipeline.
```
[the remark this answers is not in the document]
Comment (svekars) [reply examples/transactions/README.md]: added

File: doc/docs/master/deploy-manage/manage/use-transactions-to-run-multiple-commands.md [hunk 6]
```diff
@@ -0,0 +1,221 @@
...
+a collection of Pachyderm commands and execute them concurrently in a
+single job run. Regular Pachyderm operations are executed consequently,
+one after another. However, when you need to run multiple commands
+at the same time, you can use transactions.
+The transaction framework provides a method for batching together
+commit propagation such that changed branches are collected over
+the course of the transaction and all propagated in one batch at
+the end. This method allows Pachyderm to dedupe changed branches and
+branches provenant on the changed branches so that the minimum
+number of new commits are issued.
```
[the remark this answers is not in the document]
Comment (svekars) [reply doc/docs/master/deploy-manage/manage/use-transactions-to-run-multiple-commands.md]: removed

File: examples/transactions/README.md (hunk 4 above)
[the remark this answers is not in the document]
Comment (svekars) [reply examples/transactions/README.md]: removed some of the verification steps to make the procedure shorter.

File: examples/transactions/README.md [hunk 7]
```diff
@@ -0,0 +1,236 @@
...
+
+   If you check whether the pipeline has run or not, you
+   can see that it has not:
+
+   ```bash
+   $ pachctl list job --pipeline=model --no-pager
+   ID PIPELINE STARTED DURATION RESTART PROGRESS DL UL STATE
+   ```
+
+1. Add the files to the `split` repository:
```
[the remark this answers is not in the document]
Comment (svekars) [reply examples/transactions/README.md]: removed. it was a typo.

File: doc/docs/master/deploy-manage/manage/use-transactions-to-run-multiple-commands.md [hunk 8]
```diff
@@ -0,0 +1,216 @@
...
+suspends the first transaction and overwrites the transaction ID in the
+configuration file. All operations that you add to the ensuing
+transactions will be executed as soon as you close those
+transactions. To resume the initial transaction, you need to run
+`pachctl resume transaction`.
+
+Because nested transactions are not supported, transactions cannot
+conflict with each other. Every time you add a command to a transaction,
+Pachyderm creates a blueprint of the commit and verifies that the
+command is valid.
```
Comment (ysimonson) [inline doc/docs/master/deploy-manage/manage/use-transactions-to-run-multiple-commands.md]: I know this is merged already, but this isn't clear to me for a couple of reasons:

1) AFAICS, whether transactions can nest is inconsequential to whether transactions can conflict.
2) What happens if two different users are running operations in their respective transactions at the same time? How are conflicting changes resolved? (e.g. if both transactions are creating a repo with the same name?)

Comment (svekars) [reply doc/docs/master/deploy-manage/manage/use-transactions-to-run-multiple-commands.md]: @Tryneus can you comment on the above?

Comment (Tryneus) [reply doc/docs/master/deploy-manage/manage/use-transactions-to-run-multiple-commands.md]: @ysimonson 

>  whether transactions can nest is inconsequential to whether transactions can conflict

Agreed - and while transactions can't necessarily conflict (for some definition of "conflict"), one transaction can invalidate another (e.g. by deleting a branch that the other transaction references).

> What happens if two different users are running operations in their respective transactions at the same time? How are conflicting changes resolved?

Transactions are at their core a batch of operations that are applied to the cluster state in one atomic operation.  Building a transaction over multiple commands has no effect on cluster state until the transaction is committed, although each command verifies that the entire batch is still valid by doing a dry run.

If two transactions create the same repo, they will both be valid until one transaction is committed, then the other transaction will get an error when attempting to commit it or add more commands into the transaction.

**Pull request merged**